src/sonic-yang-mgmt/sonic_yang.py

```diff
@@ -0,0 +1,490 @@
+    returns:  Exception if failed
+    """
+    def add_node(self, xpath, value):
+        try:
+            node = self.new_node(xpath, value)
+            #check if the node added to the data tree
+            self.find_data_node(xpath)
+        except Exception as e:
+            print("add_node(): Failed to add data node for xpath: " + str(data_xpath))
+            self.fail(e)
+
+    """
+    merge_data(): merge a data file to the existing data tree
+    input:    yang model directory and full path of the data json file to be merged
+    returns:  Exception if failed
+    """
+    def merge_data(self, data_file, yang_dir=None):
+        #load all yang models to ctx
+        if not yang_dir:
+            yang_dir = self.yang_dir
+
+        try:
+            ctx = self.load_schema_modules_ctx(yang_dir)
+
+            #source data node
+            source_node = ctx.parse_data_path(str(data_file), ly.LYD_JSON, ly.LYD_OPT_CONFIG | ly.LYD_OPT_STRICT)
+
+            #merge
+            self.root.merge(source_node, 0)
+        except Exception as e:
+            self.fail(e)
+
+    """
+    delete_node(): delete a node from the schema/data tree
+    input:    xpath of the schema/data node
+    returns:  True - success   False - failed
+    """
+    def delete_node(self, data_xpath):
+        try:
+            node = self.find_data_node(data_xpath)
+        except Exception as e:
+            print("Failed to delete data node for xpath: " + str(data_xpath))
+            self.fail(e)
+        else:
+            if (node):
+                node.unlink()
+
+    """
+    find_node_value():  find the value of a node from the schema/data tree
+    input:    data_xpath of the data node
+    returns:  value string of the node
+    """
+    def find_node_value(self, data_xpath):
+        output = ""
+        try:
+            node = self.find_data_node(data_xpath)
+        except Exception as e:
+            print("find_node_value(): Failed to find data node from xpath: {}".format(data_xpath))
+            self.fail(e)
+        else:
+            if (node is not None):
+                subtype = node.subtype()
+                if (subtype is not None):
+                    value = subtype.value_str()
+                    return value
+            return output
+
+    """
+    set the value of a node in the data tree
+    input:    xpath of the data node
+    returns:  Exception if failed
+    """
+    def set_dnode_value(self, data_xpath, value):
+        try:
+            node = self.root.new_path(self.ctx, data_xpath, str(value), ly.LYD_ANYDATA_STRING, ly.LYD_PATH_OPT_UPDATE)
+        except Exception as e:
+            print("set data node value failed for xpath: " + str(data_xpath))
+            self.fail(e)
+
+    """
+    find_data_nodes(): find the set of nodes for the xpath
+    input:    xpath of the data node
+    returns:  list of xpath of the dataset
+    """
+    def find_data_nodes(self, data_xpath):
```

Comment:
find_nodes?

Reply:
libyang has data tree and schema tree.  This API finds data nodes, there is a separate API for find the schema node.